### PR TITLE
mClock: Add ability to handle high priority operations

### PR DESCRIPTION
**Pull request by amathuria**
There are some cases that may require mClock to handle high priority operations before other items in the queue. In order to make this possible, we are introducing a "High Queue" that will hold high priority operations. The high queue will be dequeued before the mClock queue. High queue has been implemented as a priority queue, operations will higher priority will get preference at time of dequeue.

Trello: https://trello.com/c/Kelm8z0x/775-qos-add-ability-to-handle-high-priority-operations

Signed-off-by: Aishwarya Mathuria <amathuri@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by amathuria** [comment]
jenkins test make check

**Comment by amathuria** [comment]
jenkins test make check

File: src/osd/scheduler/mClockScheduler.h [hunk 1]
```diff
@@ -130,7 +131,10 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
     true,
     true,
     2>;
+  typedef std::map<unsigned, std::list<OpSchedulerItem>, std::greater<unsigned>> SubQueue;
+  typedef SubQueue::iterator Iter;
```
Comment (ronen-fr) [inline src/osd/scheduler/mClockScheduler.h:135]: That's a too general name to create. Possible options:
1) 'using SubQueueIter = SubQueue::iterator'
2) not defining a special name for it, as on most instances you'd use 'auto'
(I'll continue reading, to see what makes sense)

File: src/osd/scheduler/mClockScheduler.cc [hunk 2]
```diff
@@ -13,6 +13,7 @@
  */
 
 
+#include <iterator>
```
Comment (ronen-fr) [inline src/osd/scheduler/mClockScheduler.cc:16]: why is this needed? are you defining a special-behaviour iterator?

File: src/osd/scheduler/mClockScheduler.cc [hunk 3]
```diff
@@ -401,6 +402,7 @@ void mClockScheduler::dump(ceph::Formatter &f) const
   // Display queue sizes
   f.open_object_section("queue_sizes");
   f.dump_int("immediate", immediate.size());
+  f.dump_int("high_queue", high_queue.size());
```
Comment (ronen-fr) [inline src/osd/scheduler/mClockScheduler.cc:405]: 'high_queue' is a terrible name. And it does not fit the others: you do not have 'immediate_queue', for example.
'priority_queue'? 'high_priority'?

File: src/osd/scheduler/mClockScheduler.h [hunk 4]
```diff
@@ -143,6 +147,18 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
     };
   }
 
+  static unsigned int get_io_prio_cut(CephContext *cct) {
+    if (cct->_conf->osd_op_queue_cut_off == "debug_random") {
+      srand(time(NULL));
```
Comment (ronen-fr) [inline src/osd/scheduler/mClockScheduler.h:152]: calling srand() every time?
and a more general question: is rand() the pref random generator for the mq? C++ has newer and better random generators.

File: src/osd/scheduler/mClockScheduler.h [hunk 5]
```diff
@@ -189,12 +205,30 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
...
+      }
+    }
+    return true;
+  }
+
   // Returns if the queue is empty
   bool empty() const final {
-    return immediate.empty() && scheduler.empty();
+    if (immediate.empty() && scheduler.empty()) {
+     if (high_queue.empty() || high_queue_empty()) {
```
Comment (ronen-fr) [inline src/osd/scheduler/mClockScheduler.h:227]: ????

File: src/osd/scheduler/mClockScheduler.h [hunk 6]
```diff
@@ -130,7 +131,10 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
     true,
     true,
     2>;
+  typedef std::map<unsigned, std::list<OpSchedulerItem>, std::greater<unsigned>> SubQueue;
```
Comment (ronen-fr) [inline src/osd/scheduler/mClockScheduler.h:134]: style note: maybe use 'using' instead of old-style 'typedef'?

File: src/osd/scheduler/mClockScheduler.cc [hunk 7]
```diff
@@ -454,17 +468,60 @@ void mClockScheduler::enqueue(OpSchedulerItem&& item)
...
+  }
   // TODO: item may not be immediate, update mclock machinery to permit
   // putting the item back in the queue
 }
 
+void mClockScheduler::enqueue_strict(unsigned priority, OpSchedulerItem&& item)
+{
+  // Enqueue the op based on priority
+  if (high_queue.find(priority) != high_queue.end()) {
+    high_queue[priority].push_front(std::move(item));
```
Comment (ronen-fr) [inline src/osd/scheduler/mClockScheduler.cc:487]: are you manually creating an std::multimap?

File: src/osd/scheduler/mClockScheduler.cc [hunk 8]
```diff
@@ -454,17 +468,61 @@ void mClockScheduler::enqueue(OpSchedulerItem&& item)
...
+  if (priority >= cutoff) {
+    immediate.push_back(std::move(item));
+  } else {
+    enqueue_strict_front(priority, std::move(item));
+  }
   // TODO: item may not be immediate, update mclock machinery to permit
   // putting the item back in the queue
 }
 
+void mClockScheduler::enqueue_strict(unsigned priority, OpSchedulerItem&& item)
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.cc:483]: Since there's a single high_priority queue, we can define a single `enqueue_strict(unsigned priority, OpSchedulerItem&& item, bool front=false)` and push_back or push_front the item into the high_priority_queue based on `front` flag.

File: src/osd/scheduler/mClockScheduler.cc [hunk 9]
```diff
@@ -480,6 +538,8 @@ WorkItem mClockScheduler::dequeue()
       return std::move(*retn.request);
     }
   }
+  // We should not reach here, must have checked empty() first
+  return {};
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.cc:542]: This can be removed as we are guaranteed to find an item from any of the queues.

File: src/osd/scheduler/mClockScheduler.h [hunk 10]
```diff
@@ -130,7 +131,12 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
     true,
     true,
     2>;
+  using SubQueue = std::map<unsigned,
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.h:134]: `using priority_t = unsigned;` and substituting in the declaration would make the purpose of the SubQueue much clearer.

File: src/osd/scheduler/mClockScheduler.h [hunk 11]
```diff
@@ -130,7 +131,12 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
...
+	std::list<OpSchedulerItem>,
+	std::greater<unsigned>>;
+  using SubQueueIter = SubQueue::iterator;
   mclock_queue_t scheduler;
+  SubQueue high_priority_queue;
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.h:139]: Using 'high_priority' for queue name is probably good enough.

File: src/osd/scheduler/mClockScheduler.h [hunk 12]
```diff
@@ -143,6 +149,18 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
     };
   }
 
+  static unsigned int get_io_prio_cut(CephContext *cct) {
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.h:152]: cct is already a private member and so you can access it without passing it to the method.

File: src/osd/scheduler/mClockScheduler.h [hunk 13]
```diff
@@ -189,12 +207,34 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
   // Enqueue the op in the front of the regular queue
   void enqueue_front(OpSchedulerItem &&item) final;
 
+  void enqueue_strict(unsigned prio, OpSchedulerItem &&item);
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.h:210]: Add appropriate comment for the enqueue_* methods and other places where new methods are introduced.

File: src/osd/scheduler/mClockScheduler.h [hunk 14]
```diff
@@ -189,12 +207,34 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
...
+      }
+    }
+    return true;
+  }
+
   // Returns if the queue is empty
   bool empty() const final {
-    return immediate.empty() && scheduler.empty();
+    if (immediate.empty() && scheduler.empty()) {
+     if (high_priority_queue_empty()) {
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.h:233]: The checks can be collapsed into a single if block.

**Review by sseshasa — CHANGES_REQUESTED** [review]
Apart from my review comments I suggest adding unit tests to test the changes. You could probably start looking at src/test/osd/TestMClockScheduler.cc.

File: src/osd/scheduler/mClockScheduler.cc [hunk 15]
```diff
@@ -454,17 +468,61 @@ void mClockScheduler::enqueue(OpSchedulerItem&& item)
 
 void mClockScheduler::enqueue_front(OpSchedulerItem&& item)
 {
-  immediate.push_back(std::move(item));
+  unsigned priority = item.get_priority();
+  unsigned cutoff = get_io_prio_cut(cct);
+
+  if (priority >= cutoff) {
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.cc:474]: In order for an item to be added to the immediate queue, the `priority >= cutoff` OR if the item is classified as op_scheduler_class::immediate. See `enqueue()` method on how to determine the classification.

File: src/osd/scheduler/mClockScheduler.h (hunk 1 above)
Comment (markhpc) [reply src/osd/scheduler/mClockScheduler.h]: To be fair, we have something of a historical precedent for using very generic names in these kinds of situations. ;)  But I do think something like SubQueueIter would be a bit more descriptive.

File: src/osd/scheduler/mClockScheduler.cc (hunk 3 above)
Comment (markhpc) [reply src/osd/scheduler/mClockScheduler.cc]: I don't hate it, but I agree they should follow the same convention whatever is chosen.

File: src/osd/scheduler/mClockScheduler.cc (hunk 8 above)
Comment (amathuria) [reply src/osd/scheduler/mClockScheduler.cc:483]: That makes sense, I'll change it.

File: src/osd/scheduler/mClockScheduler.cc (hunk 9 above)
Comment (amathuria) [reply src/osd/scheduler/mClockScheduler.cc:542]: I think one of the unit tests had reached here because it didn't check empty() first. We never dequeue without checking empty() in other places. Should I modify the test?

Comment (sseshasa) [reply src/osd/scheduler/mClockScheduler.cc:542]: In the tests we always try to match the number of enqueue and dequeue operations. If the code reaches this point when you expect an item to be dequeued, then it probably indicates something amiss in the enqueue or dequeue logic.

File: src/osd/scheduler/mClockScheduler.h [hunk 16]
```diff
@@ -186,15 +206,35 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
...
+	return false;
+      }
+    }
+    return true;
+  }
+
   // Returns if the queue is empty
   bool empty() const final {
-    return immediate.empty() && scheduler.empty();
+    if (immediate.empty() && scheduler.empty()
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.h:233]: `return immediate.empty() && scheduler.empty() && high_priority_empty();`

You could just do with the above in one line.

File: src/osd/scheduler/mClockScheduler.h [hunk 17]
```diff
@@ -130,7 +131,13 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
     true,
     true,
     2>;
+  using priority_t = unsigned;
+  using SubQueue = std::map<priority_t,
+	std::list<OpSchedulerItem>,
+	std::greater<unsigned>>;
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.h:137]: std::greater<priority_t>

File: src/test/osd/TestMClockScheduler.cc [hunk 18]
```diff
@@ -169,3 +181,31 @@ TEST_F(mClockSchedulerTest, TestMultiClientOrderedEnqueueDequeue) {
   }
   ASSERT_TRUE(q.empty());
 }
+
+TEST_F(mClockSchedulerTest, TestHighPriorityQueueEnqueueDequeue) {
```
Comment (sseshasa) [inline src/test/osd/TestMClockScheduler.cc:185]: Test looks good. It would be good to have additional test(s) that creates a mix of immediate, high_priority and mclock scheduler items and test if they get dequeued in the expected way. The idea is ensure maximum coverage possible of the code that is introduced.

File: src/osd/scheduler/mClockScheduler.h (hunk 12 above)
Comment (sseshasa) [reply src/osd/scheduler/mClockScheduler.h:153]: You can disregard the above comment. We can let this function be static.

File: src/osd/scheduler/mClockScheduler.h [hunk 19]
```diff
@@ -186,15 +206,31 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
...
+  // Enqueue the op in the front of the high priority queue or the immediate queue (based on priority)
   void enqueue_front(OpSchedulerItem &&item) final;
 
+  // Enqueue the op to the high priority queue
+  void enqueue_strict(unsigned prio, OpSchedulerItem &&item, bool front = false);
+
   // Return an op to be dispatch
   WorkItem dequeue() final;
 
+  bool high_priority_empty() const {
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.h:218]: nit: Add appropriate comment for the high_priority_empty() method to be consistent.

File: src/test/osd/TestMClockScheduler.cc [hunk 20]
```diff
@@ -169,3 +181,74 @@ TEST_F(mClockSchedulerTest, TestMultiClientOrderedEnqueueDequeue) {
...
+  ASSERT_EQ(201u, r.get_map_epoch());
+
+  r = get_item(q.dequeue());
+  ASSERT_EQ(200u, r.get_map_epoch());
+
+  ASSERT_TRUE(q.high_priority_empty());
+}
+
+TEST_F(mClockSchedulerTest, TestAllQueuesEnqueueDequeue) {
+	ASSERT_TRUE(q.empty());
```
Comment (sseshasa) [inline src/test/osd/TestMClockScheduler.cc:214]: nit: Indentation is off.

File: src/test/osd/TestMClockScheduler.cc [hunk 21]
```diff
@@ -169,3 +181,74 @@ TEST_F(mClockSchedulerTest, TestMultiClientOrderedEnqueueDequeue) {
...
+  }
+
+  ASSERT_FALSE(q.empty());
+  auto r = get_item(q.dequeue());
+  // Immediate queue should be dequeued first
+  ASSERT_EQ(103u, r.get_map_epoch());
+  r = get_item(q.dequeue());
+  ASSERT_EQ(104u, r.get_map_epoch());
+
+  // High queue should be dequeuend second
```
Comment (sseshasa) [inline src/test/osd/TestMClockScheduler.cc:241]: s/dequeuend/dequeued

File: src/test/osd/TestMClockScheduler.cc [hunk 22]
```diff
@@ -169,3 +181,74 @@ TEST_F(mClockSchedulerTest, TestMultiClientOrderedEnqueueDequeue) {
...
+  // higher priority operation first
+  r = get_item(q.dequeue());
+  ASSERT_EQ(201u, r.get_map_epoch());
+  r = get_item(q.dequeue());
+  ASSERT_EQ(200u, r.get_map_epoch());
+
```
Comment (sseshasa) [inline src/test/osd/TestMClockScheduler.cc:247]: Add comment here for the mClock queue.

File: src/osd/scheduler/mClockScheduler.cc [hunk 23]
```diff
@@ -454,17 +468,63 @@ void mClockScheduler::enqueue(OpSchedulerItem&& item)
...
+    enqueue_strict(priority, std::move(item), true);
+  }
   // TODO: item may not be immediate, update mclock machinery to permit
   // putting the item back in the queue
 }
 
+void mClockScheduler::enqueue_strict(unsigned priority,
+                                     OpSchedulerItem&& item,
+				     bool front)
+{
```
Comment (athanatos) [inline src/osd/scheduler/mClockScheduler.cc:487]: std::map<>::operator[] will default construct entries that don't exist, no need to check:

```
void mClockScheduler::enqueue_strict(unsigned priority,
                                     OpSchedulerItem&& item,
				     bool front)
{
  if (front) {
    high_priority[priority].push_front(std::move(item));
  } else {
    high_priority[priority].push_back(std::move(item));
  }
}
```

File: src/osd/scheduler/mClockScheduler.cc [hunk 24]
```diff
@@ -454,17 +468,63 @@ void mClockScheduler::enqueue(OpSchedulerItem&& item)
 
 void mClockScheduler::enqueue_front(OpSchedulerItem&& item)
 {
-  immediate.push_back(std::move(item));
+  unsigned priority = item.get_priority();
+  unsigned cutoff = get_io_prio_cut(cct);
+  auto id = get_scheduler_id(item);
+
```
Comment (athanatos) [inline src/osd/scheduler/mClockScheduler.cc:474]: This is a bit odd -- ```priority >= cutoff```` but ```op_scheduler_class::immediate != id.class_id``` in enqueue would cause the op to be put into the high queue rather than the immediate queue.

File: src/osd/scheduler/mClockScheduler.cc [hunk 25]
```diff
@@ -454,17 +468,63 @@ void mClockScheduler::enqueue(OpSchedulerItem&& item)
...
+  if (priority >= cutoff || op_scheduler_class::immediate == id.class_id) {
+    immediate.push_back(std::move(item));
+  } else {
+    enqueue_strict(priority, std::move(item), true);
+  }
```
Comment (athanatos) [inline src/osd/scheduler/mClockScheduler.cc:479]: How about we do the following instead:

```

void mClockScheduler::enqueue_front(OpSchedulerItem&& item)
{
  unsigned priority = item.get_priority();
  unsigned cutoff = get_io_prio_cut(cct);
  auto id = get_scheduler_id(item);

  if (op_scheduler_class::immediate == id.class_id) {
    immediate.push_back(std::move(item));
  } else if (priority >= cutoff) {
    enqueue_strict(priority, std::move(item), true);
  } else {
    // mclock doesn't support enqueue at front, so we use the high queue
    // with priority 0
    enqueue_strict(0, std::move(item), true);
  }
}
```

File: src/osd/scheduler/mClockScheduler.cc (hunk 23 above)
Comment (athanatos) [reply src/osd/scheduler/mClockScheduler.cc:487]: Also, enqueue_strict places items into the high_priority queue -- let's call it enqueue_high.

File: src/osd/scheduler/mClockScheduler.h [hunk 26]
```diff
@@ -186,15 +206,32 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
   // Enqueue op in the back of the regular queue
   void enqueue(OpSchedulerItem &&item) final;
 
-  // Enqueue the op in the front of the regular queue
+  // Enqueue the op in the front of the high priority queue or the immediate queue (based on priority)
   void enqueue_front(OpSchedulerItem &&item) final;
 
+  // Enqueue the op to the high priority queue
+  void enqueue_strict(unsigned prio, OpSchedulerItem &&item, bool front = false);
```
Comment (athanatos) [inline src/osd/scheduler/mClockScheduler.h:213]: This should be private.

File: src/osd/scheduler/mClockScheduler.cc [hunk 27]
```diff
@@ -454,17 +468,63 @@ void mClockScheduler::enqueue(OpSchedulerItem&& item)
...
+	++next_it;
+	high_priority.erase(it->first);
+	continue;
+      }
+      break;
+    }
+    WorkItem work_item{std::move(it->second.back())};
+    it->second.pop_back();
+    ceph_assert(std::get_if<OpSchedulerItem>(&work_item));
+    return work_item;
```
Comment (athanatos) [inline src/osd/scheduler/mClockScheduler.cc:527]: Let's be strict about this: high_priority entries are *never* empty.  Thus:

```
  } else if (!high_priority.empty()) {
    auto iter = high_priority.begin();
    // invariant: high_priority entries are never empty
    assert(!iter->second.empty());
    WorkItem ret{std::move(iter->second.back())};
    iter->second.pop_back();
    if (iter->second.empty()) {
      // maintain invariant, high priority entries are never empty
      high_priority.erase(iter);
    }
    ceph_assert(std::get_if<OpSchedulerItem>(&ret));
    return ret;
  } else {
```

File: src/osd/scheduler/mClockScheduler.h [hunk 28]
```diff
@@ -130,7 +131,13 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
     true,
     true,
     2>;
+  using priority_t = unsigned;
+  using SubQueue = std::map<priority_t,
+	std::list<OpSchedulerItem>,
+	std::greater<priority_t>>;
+  using SubQueueIter = SubQueue::iterator;
   mclock_queue_t scheduler;
+  SubQueue high_priority;
```
Comment (athanatos) [inline src/osd/scheduler/mClockScheduler.h:140]:
```
  /**
   * high_priority
   *
   * Holds entries to be dequeued in strict order ahead of mclock.
   * Invariant: entries are never empty
   */
```

Comment (athanatos) [reply src/osd/scheduler/mClockScheduler.h:140]: Also, it seems to me like we can remove immediate and use a special priority value of std::numeric_limits<priority_t>::max().

File: src/osd/scheduler/mClockScheduler.h [hunk 29]
```diff
@@ -186,15 +206,32 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
...
 
-  // Enqueue the op in the front of the regular queue
+  // Enqueue the op in the front of the high priority queue or the immediate queue (based on priority)
   void enqueue_front(OpSchedulerItem &&item) final;
 
+  // Enqueue the op to the high priority queue
+  void enqueue_strict(unsigned prio, OpSchedulerItem &&item, bool front = false);
+
   // Return an op to be dispatch
   WorkItem dequeue() final;
```
Comment (athanatos) [inline src/osd/scheduler/mClockScheduler.h:217]: Because high_priority items are never empty, this is unnecessary, simply use high_priority.empty().

File: src/osd/scheduler/mClockScheduler.h (hunk 4 above)
Comment (athanatos) [reply src/osd/scheduler/mClockScheduler.h]: Yup, let's use std::random rather than srand.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/osd/scheduler/mClockScheduler.h (hunk 28 above)
Comment (amathuria) [reply src/osd/scheduler/mClockScheduler.h:146]: @athanatos that makes sense to me. Do you think I can do it as part of this change?

File: src/osd/scheduler/mClockScheduler.h [hunk 30]
```diff
@@ -130,8 +131,20 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
...
   mclock_queue_t scheduler;
-  std::list<OpSchedulerItem> immediate;
+  /**
+   * high_priority
+   *
+   * Holds entries to be dequeued in strict order ahead of mClock
+   * Invariant: entries are never empty
+   */
+  SubQueue high_priority;
+  priority_t immediate_class_priority = 999999;
```
Comment (athanatos) [inline src/osd/scheduler/mClockScheduler.h:147]: Use std::numeric_limits<priority_t>::max()

**Review by athanatos — APPROVED** [review]
LGTM other than using numeric_limits rather than a hardcoded value.

File: src/osd/scheduler/mClockScheduler.h [hunk 31]
```diff
@@ -130,8 +131,20 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
     true,
     true,
     2>;
+  using priority_t = unsigned;
+  using SubQueue = std::map<priority_t,
+	std::list<OpSchedulerItem>,
+	std::greater<priority_t>>;
+  using SubQueueIter = SubQueue::iterator;
```
Comment (sseshasa) [inline src/osd/scheduler/mClockScheduler.h:138]: SubQueueIter is not used and can be removed.

**Review by sseshasa — APPROVED** [review]
LGTM! Let's get the rados run going.

**Comment by amathuria** [comment]
http://pulpito.front.sepia.ceph.com/amathuri-2023-02-08_09:43:45-rados-main-distro-default-smithi/

File: src/osd/scheduler/mClockScheduler.h (hunk 30 above)
Comment (amathuria) [reply src/osd/scheduler/mClockScheduler.h]: Done

**Comment by amathuria** [comment]
jenkins test make check

**Comment by amathuria** [comment]
jenkins test api

**Comment by ljflores** [comment]
Rados suite review:
http://pulpito.front.sepia.ceph.com:80/yuriw-2023-02-22_21:31:50-rados-wip-yuri10-testing-2023-02-22-0848-distro-default-smithi
http://pulpito.front.sepia.ceph.com:80/yuriw-2023-02-23_16:14:52-rados-wip-yuri10-testing-2023-02-22-0848-distro-default-smithi

Failures, unrelated:
    1. https://tracker.ceph.com/issues/57754
    2. https://tracker.ceph.com/issues/58475
    3. https://tracker.ceph.com/issues/58585
    4. https://tracker.ceph.com/issues/58797

Details:
    1. test_envlibrados_for_rocksdb.sh: update-alternatives: error: alternative path /usr/bin/gcc-11 doesn't exist - Ceph - RADOS
    2. test_dashboard_e2e.sh: Conflicting peer dependency: postcss@8.4.21 - Ceph - Mgr - Dashboard
    3. rook: failed to pull kubelet image - Ceph - Orchestrator
    4. scrub/osd-scrub-dump.sh: TEST_recover_unexpected fails from "ERROR: Unexpectedly low amount of scrub reservations seen during test"

Some failures in the runs were directly related to https://github.com/ceph/ceph/pull/49837.
